### ISSUE TITLE
Is there a way to find the probability or confidence score of the extracted named entities?

### ISSUE THREAD

**Issue by faizan30** [explosion/spaCy#831]
Usage

## Your Environment

* Operating System: Windows 10
* Python Version Used: 2.7, 64 bit
* spaCy Version Used: 1.6.0
* Environment Information: Anaconda

**Comment by honnibal** [comment]
The entity recogniser uses a transition-based model with greedy decoding. On every word it decides between the following actions:

* **Begin**: begin a new entity
* **In:** Continue the current entity
* **Last**: End the current entity (adding this token to it)
* **Unit**: Mark this work as a single-word entity
* **Out**: Mark the current word as not an entity.

The model scores each action on every word, unless only one action is available, e.g. at the end of a sentence. In that situation.

You can get the scores with the following method, although it's quite inefficient:

```python

scores = numpy.zeros((len(doc), nlp.entity.model.nr_class))
with nlp.entity.step_through(doc) as state:
    while not state.is_final:
        action = state.predict()
        next_tokens = state.queue
        scores[next_tokens[0].i] = state.scores
        state.transition(action)
```

You'll want to apply a softmax transformation over the raw scores so you can work with them as probabilities.

The model is trained on 0/1 loss using the perceptron algorithm, and the feature set is sparse and redundant. This leads to quite poorly calibrated scores.

The per-word scores also don't really answer the question we're interested in. The subsequent scores are conditioned on the previous analysis. Let's say we make a mistake at token 7, and by token 10 we get enough evidence to see this clearly. The model at token 10 is not going to say, "Oh, I'm confused". It's going to say, "what's the best thing to do here?". The decision at token 7 has been made, and those costs are sunk. And after the bad decision has been made, the least-bad follow up actions are often really certain.

Basically: everything about the design of the model is geared towards getting out the best 1-best sequence. If you need ambiguity, there's a beam-search implementation in the beam branch that you might want to look at. If you can get that running, you'll have much better luck getting alternative analyses.

**Comment by honnibal** [comment]
Follow #881 for progress on this.

**Comment by faizan86** [comment]
Thank you.

**Comment by rbhambriiit** [comment]
The code above seems bit out-of-date

![image](https://user-images.githubusercontent.com/13808100/48684742-16568680-ebd9-11e8-9418-b657f1c0ea87.png)


nlp.entity.model returns a tuple for me:
(<thinc.api.FeedForward at 0x7f8eaeb30898>,
 <spacy._ml.PrecomputableAffine at 0x7f8eaeb3f470>,
 <thinc.api.FeedForward at 0x7f8eaeb3a080>)


I imagine one of these used to be returned earlier.

Also none of these 3 has a method/attribute for nr_class

**Comment by svlandeg** [comment]
Merging with the more recent thread https://github.com/explosion/spaCy/issues/5917
